### PR TITLE
Ticket 3 GET /api/topics done

### PR DESCRIPTION
**Pull request by xtej4ik**

### CODE REVIEW

File: models/topics.models.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+exports.fetchTopics = () =>{
+    
+    return db.query(`SELECT * FROM topics;`)
+    .then((result) => {
+       return result.rows;
+      });
+}
+
+
+exports.addTopics = (topicAdd) => {
```
Comment (B7Bprog) [inline models/topics.models.js]: This shouldn't be part of this pull request.

File: controllers/topics.controllers.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+const { fetchTopics } = require("../models/topics.models");
+
+exports.getTopics = (req, res, next) => {
+  fetchTopics()
+    .then((output) => {
+      res.status(200).send({ topics: output });
+    })
```
Comment (B7Bprog) [inline controllers/topics.controllers.js]: Please include a .catch() block here.

File: __tests__/app.test.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+             slug: expect.any(String),
+            });
+          });
+        });
+    });
+    describe('GET/not-a-route', () => {
+      it('responds with status 404', () => {
+          return request(app)
+          .get('/not-a-route')
+          .expect(404)
```
Comment (B7Bprog) [inline __tests__/app.test.js:49]: Please check for error message as well here.

File: app.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+const app = express();
+
+app.use(express.json())
+
+app.get('/api', (req, res) => {
+    res.status(200).send({msg: 'Server is up and running.'})
+});
+
+app.get('/api/topics', getTopics);
+
```
Comment (B7Bprog) [inline app.js:14]: At the moment the 404 error seems to come from 'express'. Please take control and use app.all('*',()=>{}) to catch all wrong path. 

https://www.geeksforgeeks.org/express-js-app-all-function/

File: app.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+
```
Comment (B7Bprog) [inline app.js:25]: Please include an error handling middleware for the 500 server errors.

**Review by B7Bprog — CHANGES_REQUESTED** [review]
Hey Inna, please read my comments, and refactor your code, then send us a quick pull request with it.

File: controllers/topics.controllers.js [hunk 6]
```diff
@@ -5,4 +5,8 @@ exports.getTopics = (req, res, next) => {
     .then((output) => {
       res.status(200).send({ topics: output });
     })
+    .catch((err) => {
+      console.err(err);
+      res.status(500).send({ err: "Unable to retrieve topics!" });
```
Comment (HeatherEver) [inline controllers/topics.controllers.js]: the 500 error should not be in the .catch block this should be an external error handler in app.js
https://notes.northcoders.com/courses/js-back-end/error-handling
see the second example in Error Handling Middleware in the notes to refer to 😊

File: __tests__/app.test.js [hunk 7]
```diff
@@ -45,7 +45,8 @@ describe("GET/api/", () => {
       it('responds with status 404', () => {
           return request(app)
           .get('/not-a-route')
-          .expect(404)
+          .expect(404).send({ msg: 'Route not found' });
+
```
Comment (HeatherEver) [inline __tests__/app.test.js]: we don't want a `.send` here, you want to be testing what comes back from your error handler so you can test the message on the body of the error 
For example
```
test('status:400, responds with an error message when passed a bad user ID', () => {
  return request(app)
    .get('/api/users/notAnID')
    .expect(400)
    .then(({ body }) => {
      expect(body.msg).toBe('Invalid input');
    });
});
// ...
```
(taken from https://notes.northcoders.com/courses/js-back-end/error-handling)

File: app.js [hunk 8]
```diff
@@ -15,6 +13,10 @@ app.get('/api', (req, res) => {
 app.get('/api/topics', getTopics);
 
 
+// to catch all wrong path
+app.all('*', (req, res) => {
```
Comment (HeatherEver) [inline app.js]: might want to add a `/` before the `*` so it's `/*`

File: controllers/topics.controllers.js (hunk 6 above)
Comment (HeatherEver) [reply controllers/topics.controllers.js]: and then think about what we could use in this catch block to pass the error along....

**Review by HeatherEver — COMMENTED** [review]
Nice changes so far, just a couple more to make

File: app.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+const express = require('express')
+
+const { getTopics } = require("./controllers/topics.controllers")
+
+const app = express();
+
+app.use(express.json())
```
Comment (Poonam-raj) [inline app.js:7]: Ideally we would not include the express.json line until you have a body that you need to use in a endpoint.

File: controllers/topics.controllers.js (hunk 6 above)
Comment (Poonam-raj) [reply controllers/topics.controllers.js]: Again, Inna, the 500 error handler should be in app.js as per the linked notes. make this change before you merge

**Review by Poonam-raj — APPROVED** [review]
Hey Inna, thanks for making changes!

I want you to amend the 500 error handler so it exists in your app.ts file instead as per the notes Heather has linked.

Once you have made this change I am happy for you to merge 🎉
